app/models/manageiq/providers/vmware/cloud_manager/refresh_parser.rb

```diff
@@ -176,4 +178,32 @@ def parse_image(image, is_public)
 
     return uid, new_result
   end
+
+  def parse_vapp_template(vapp_template)
+    uid = vapp_template.id
+
+    begin
+      # The content of the template is the OVF specification of the vApp template
+      content = @connection.get_vapp_template_ovf_descriptor(uid).body
+
+      new_result = {
+        :type        => ManageIQ::Providers::Vmware::CloudManager::OrchestrationTemplate.name,
+        :ems_ref     => uid,
+        :name        => vapp_template.name,
+        :description => vapp_template.description,
+        :orderable   => false,
+        :content     => content,
+        # By default #save_orchestration_templates_inventory does not set the EMS
+        # ID because templates are not EMS specific. We are setting the EMS
+        # explicitly here, because vapps are specific to concrete EMS.
+        :ems_id      => @ems.id
+      }
+
+      return uid, new_result
+    rescue
+      # Without content we should not create the orchestration template because
+      # these are identified by it (or its MD5)
+      $log.info("MIQ(#{self.class.name}.#{__method__}) Downloading vApp template for vApp template #{uid} failed")
```

Comment:
@agrare Basically, the only exceptions that can occur are related to either connectivity error or unauthorised or forbidden access. The former can occur anywhere during the refresh so it doesn't make sense to add this specific check here. I don't think other providers are dealing with them either.

Unauthorised access means that the EMS auth is not properly configured, however this can only occur if it was changed during the refresh -- failing entire refresh is ok in this case.

The only interesting might therefore be the "Forbidden", however this is again only possible if access has been revoked between downloading the vApp template and requesting to get OVF descriptor. Again, if we fail the entire refresh, this will be remedied during the next refresh when the vApp template will not even be present.

I have added a sample code (below) that could be used to explain in more details what is going on. This code actually returns an empty result after receiving Forbidden error.

```ruby
    # The content of the template is the OVF specification of the vApp template
    begin
      content = @connection.get_vapp_template_ovf_descriptor(uid).body
    rescue Fog::Compute::VcloudDirector::Forbidden => err
      $log.warn "Forbidden to read OVF descriptor of the vApp template #{uid} in provider #{ems.name} (#{ems.id})."\
                "Message=#{err.message}"

      # If it is forbidden to read the OVF, the vApp template should not be saved into inventory. Thus, return
      # empty result.
      return
    rescue => err
      $log.error "Unable to obtain OVF descriptor of the vApp template #{uid} in provider #{ems.name} (#{ems.id})."\
                "Message=#{err.message}"

      # Cancel the refresh by raising new exception. This can be caused either by network problem or inpromer
      # configuration of the EMS provider or VMware installation.
      raise MiqException::MiqVmwareCloudApiRequestError,
            "Unable to obtain OVF descriptor of the vApp template #{uid} in provider #{ems.name} (#{ems.id}) "\
            "through API. Please, fix your VMware installation and run refresh again."
    end
```

I can go either way. Even without exceptions I believe we are safe. What do you think?